metron-platform/metron-common/src/main/java/org/apache/metron/common/dsl/functions/resolver/BaseFunctionResolver.java

```diff
@@ -0,0 +1,198 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.metron.common.dsl.functions.resolver;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import org.apache.commons.lang.ObjectUtils;
+import org.apache.metron.common.dsl.Context;
+import org.apache.metron.common.dsl.Stellar;
+import org.apache.metron.common.dsl.StellarFunction;
+import org.apache.metron.common.dsl.StellarFunctionInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * The base implementation of a function resolver.
+ *
+ */
+public abstract class BaseFunctionResolver implements FunctionResolver, Serializable {
+
+  protected static final Logger LOG = LoggerFactory.getLogger(BaseFunctionResolver.class);
+
+  /**
+   * Maps a function name to the metadata necessary to execute the Stellar function.
+   */
+  private Supplier<Map<String, StellarFunctionInfo>> functions;
+
+  /**
+   * The Stellar execution context that can be used to inform the function resolution process.
+   */
+  protected Context context;
+
+  public BaseFunctionResolver() {
+    // memoize provides lazy initialization and thread-safety (the ugly cast is necessary for serialization)
+    functions = Suppliers.memoize((Supplier<Map<String, StellarFunctionInfo>> & Serializable) this::resolveFunctions);
+  }
```

Comment:
If a function scan is currently running, then competing threads will block when calling `function.get()...`